lighthouse-core/gather/gatherers/source-maps.js

```diff
@@ -0,0 +1,118 @@
+
+/* global fetch */
+
+/**
+ * This function is careful not to parse the response as JSON, as it will
+ * just need to be serialized again over the protocol, and source maps can
+ * be huge.
+ *
+ * @param {string} url
+ */
+/* istanbul ignore next */
+async function fetchSourceMap(url) {
+  try {
+    const response = await fetch(url);
+    return response.text();
+  } catch (err) {
+    return {errorMessage: err.toString()};
+  }
+}
+
+/**
+ * @fileoverview Gets JavaScript source maps.
+ */
+class SourceMaps extends Gatherer {
+  constructor() {
+    super();
+    /** @type {LH.Crdp.Debugger.ScriptParsedEvent[]} */
+    this._scriptParsedEvents = [];
+    this.onScriptParsed = this.onScriptParsed.bind(this);
+  }
+
+  /**
+   * @param {Driver} driver
+   * @param {string} sourceMapUrl
+   * @return {Promise<{map: import('source-map').RawSourceMap} | {errorMessage: string}>}
+   */
+  async fetchSourceMapInPage(driver, sourceMapUrl) {
+    // TODO: change default protocol timeout?
+    // driver.setNextProtocolTimeout(250);
+    /** @type {string|{errorMessage: string}} */
+    const sourceMapJson =
+      await driver.evaluateAsync(`(${fetchSourceMap})(${JSON.stringify(sourceMapUrl)})`);
+
+    if (typeof sourceMapJson === 'object') {
+      return sourceMapJson;
+    }
+
+    return {map: JSON.parse(sourceMapJson)};
+  }
+
+  /**
+   * @param {LH.Crdp.Debugger.ScriptParsedEvent} event
+   */
+  onScriptParsed(event) {
+    if (event.sourceMapURL) {
+      this._scriptParsedEvents.push(event);
+    }
+  }
+
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   */
+  async beforePass(passContext) {
+    const driver = passContext.driver;
+    driver.on('Debugger.scriptParsed', this.onScriptParsed);
+    await driver.sendCommand('Debugger.enable');
+  }
+
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   * @return {Promise<LH.Artifacts['SourceMaps']>}
+   */
+  async afterPass(passContext) {
+    const driver = passContext.driver;
+
+    driver.off('Debugger.scriptParsed', this.onScriptParsed);
+    await driver.sendCommand('Debugger.disable');
+
+    /** @type {LH.Artifacts.SourceMap[]} */
+    const sourceMaps = [];
+    for (const event of this._scriptParsedEvents) {
+      if (!event.sourceMapURL) continue;
+
+      const url = event.url;
+      try {
+        if (event.sourceMapURL.startsWith('data:')) {
+          const buffer = Buffer.from(event.sourceMapURL.split(',')[1], 'base64');
+          sourceMaps.push({
+            url,
+            map: JSON.parse(buffer.toString()),
+          });
+        } else {
+          const fetchedSourceMap = await this.fetchSourceMapInPage(driver, event.sourceMapURL);
+          sourceMaps.push({
+            url,
+            ...fetchedSourceMap,
+          });
+        }
+      } catch (err) {
+        // Without this catch, this silently fails and the gatherer returns an empty object... why no visible error?
```

Comment:
is that for all errors or just protocol timeout LHError? only idea I can think of is something to do with an `Error` being stringified at some point which tends to result in the empty object

Reply:
perhaps it was the error thing - at one point I wasn't wrapping the "page function" code to return `{error: ...}`, so perhaps whatever error was thrown just bubbled up through the protocol as an empty object ..

Reply:
so is the try/catch still necessary? Ideally, yeah, the `evaluateAsync` is isolated so we shouldn't need it

Reply:
I'll reduce it to where it's necessary.

But, yeah this is still weird, and applies to any gatherer. If a gatherer throws an error, the resulting artifact is just an empty object.

a gatherer's afterPass
```
throw new Error('blasjdlkasj');
```

artifacts.json
```
"SourceMaps": {},
```

I don't even see an error in the terminal.

Reply:
Oh, I see. Yes, that's the unfortunate effect of `JSON.stringify(new Error('blasjdlkasj'))`

bug is #4984